tensorflow-tools/src/main/java/org/tensorflow/tools/buffer/DataStorageVisitor.java

```diff
@@ -0,0 +1,147 @@
+   *
+   * @param buffer underlying buffer
+   * @return any value
+   * @see DataBuffer#accept(DataStorageVisitor)
+   */
+  default R visit(LongBuffer buffer) {
+    return otherwise();
+  }
+
+  /**
+   * Visit the {@link FloatBuffer} backing a given instance of a {@link DataBuffer}
+   *
+   * @param buffer underlying buffer
+   * @return any value
+   * @see DataBuffer#accept(DataStorageVisitor)
+   */
+  default R visit(FloatBuffer buffer) {
+    return otherwise();
+  }
+
+  /**
+   * Visit the {@link DoubleBuffer} backing a given instance of a {@link DataBuffer}
+   *
+   * @param buffer underlying buffer
+   * @return any value
+   * @see DataBuffer#accept(DataStorageVisitor)
+   */
+  default R visit(DoubleBuffer buffer) {
+    return otherwise();
+  }
+
+  /**
+   * Visit the boolean array backing a given instance of a {@link DataBuffer}
+   *
+   * @param array underlying array
+   * @param offset offset of the buffer within the array
+   * @param length length of the buffer within the array
+   * @return any value
+   * @see DataBuffer#accept(DataStorageVisitor)
+   */
+  default R visit(boolean[] array, int offset, int length) {
+    return otherwise();
+  }
+
+  /**
+   * Visit the bit set backing a given instance of a {@link DataBuffer}
+   *
+   * @param bitSet underlying bit set
+   * @param offset offset of the buffer within the bit set
+   * @param numBits number of bits used to represent the buffer within the bit set
+   * @return any value
+   * @see DataBuffer#accept(DataStorageVisitor)
+   */
+  default R visit(BitSet bitSet, int offset, long numBits) {
+    return otherwise();
+  }
+
+  /**
+   * Visit the object array backing a given instance of a {@link DataBuffer}
+   *
+   * @param array underlying array
+   * @param offset offset of the buffer within the array
+   * @param length length of the buffer within the array
+   * @return any value
+   * @see DataBuffer#accept(DataStorageVisitor)
+   */
+  default R visit(Object[] array, int offset, int length) {
+    return otherwise();
+  }
+
+  /**
+   * Visit the raw memory segment of a given instance of a {@link DataBuffer}
+   *
+   * @param address native address of the buffer
+   * @param length length of the buffer
+   * @param scale number of bytes required to store a single value of this buffer
+   * @return any value
+   * @see DataBuffer#accept(DataStorageVisitor)
+   */
+  default R visit(long address, long length, long scale) {
+    return otherwise();
+  }
+
+  /**
+   * Fallback method called if the visitor implementation does not support the type of backing storage
+   * for a given {@link DataBuffer}
+   *
+   * <p>The implementor of this interface must override the {@code visit} methods for type of storage
+   * it supports. If {@link DataBuffer#accept(DataStorageVisitor)} is called on a buffer
+   * using a different type of storage, the invocation will fallback to this method.
+   *
+   * @return any value
+   */
+  R otherwise();
```

Comment:
This method is called when the visitor does not handle a given type of data storage, so that it can fallback to a generic task if needed. I still prefer `otherwise()` to `defaultVisit()` because basically when this method is called, it means that you are not visiting any data storage. I used it for example like this:

```java
    return otherBuffer.accept(new DataStorageVisitor<Boolean>() {

      @Override
      public Boolean visit(DoubleBuffer otherStorage) {
        return myStorage.equals(otherStorage);
      }

      @Override
      public Boolean otherwise() {
        for (int idx = 0; idx < size(); ++idx) {
          if (otherBuffer.getDouble(idx) != getDouble(idx)) {
            return false;
          }
        }
        return true;
      }
    });
```

So when `otherwise()` is call, I can only build a routine using the public API of a `DataBuffer` and nothing internal like other `visit()` methods can do.

Reply:
I'd still consider that to be visiting every index in the storage, but I don't write that many visitors to be honest. Maybe `fallback`? The name `otherwise` to me suggests there should be some condition that makes it go down that path that the user could control, or is data dependent. The use case here seems to be type dependent (in that it'll be chosen statically at compile time).

Reply:
Ok no problem for "fallback", as long as it is not "visit" I'm fine.